modules/aws/s3.go

```diff
@@ -260,6 +263,81 @@ func EmptyS3BucketE(t testing.TestingT, region string, name string) error {
 	return err
 }
 
+// PutS3BucketLogging enables logging for the given S3 bucket in the given region
+// Make sure that the logsTargetBucket exists prior to using it here.
+func PutS3BucketLogging(t testing.TestingT, region string, bucketName string, logsTargetBucketName string) {
+	err := PutS3BucketLoggingE(t, region, bucketName, logsTargetBucketName)
+	require.NoError(t, err)
+}
+
+// PutS3BucketLoggingE enables logging for the given S3 bucket in the given region
+// Make sure that the logsTargetBucket exists prior to using it here.
+func PutS3BucketLoggingE(t testing.TestingT, region string, sourceBucketName string, logsTargetBucketName string) error {
+	logger.Logf(t, "Applying logging to S3 bucket %s in %s", sourceBucketName, region)
+	logger.Logf(t, "Setting TargetBucket to be %s", logsTargetBucketName)
+
+	s3Client, err := NewS3ClientE(t, region)
+	if err != nil {
+		return err
+	}
+
+	logsUri := fmt.Sprintf("uri=%s", logDeliveryUri)
+
+	//Put ACL permissions on the targetBucket, as it's different from the origin S3 bucket
+	aclInput := s3.PutBucketAclInput{
+		Bucket:       aws.String(logsTargetBucketName),
+		GrantWrite:   aws.String(logsUri),
+		GrantReadACP: aws.String(logsUri),
+	}
+
+	if _, err := s3Client.PutBucketAcl(&aclInput); err != nil {
+		return err
+	}
+
+	//enable logging on the origin S3 bucket
+	params := &s3.PutBucketLoggingInput{
+		Bucket: aws.String(sourceBucketName),
+		BucketLoggingStatus: &s3.BucketLoggingStatus{
+			LoggingEnabled: &s3.LoggingEnabled{
+				TargetBucket: aws.String(logsTargetBucketName),
+				TargetPrefix: aws.String("/"),
+			},
+		},
+	}
+
+	_, err = s3Client.PutBucketLogging(params)
+	return err
+}
```

Comment:
Can you share what use case prompted you to add the put version? The reason I ask is that this technically doesn't pass [the litmus test](https://terratest.gruntwork.io/docs/community/contributing/#infrastructure-and-validation-helpers) for contributions because it is an API for modifications, something that isn't really necessary in a testing library.

Reply:
Yeah, that's a very good question @yorinasub17 

Initially, my thinking was "don't overclutter terratest, and just add the code I needed" _- as in the `GetLoggingTargetBucketName function`_

As I got to this yesterday though, I felt uncomfortable adding in code that was not tested, which then prompted me to add the PutBucketLogging method.

It wasn't a lot of extra effort, so I did add it in the end. Do you think it's smarter at this point to maybe skip that testing bit entirely? Or am I missing an obvious way how I can test without the prior setup? 

As it was a difficult trade-off, I thought it will be safer to push code that is tested. But as this was my main wondering _(and the reason why I didn't add the tests initially when the PR was opened)_, I'm very happy to take guidance and learn from this!

Reply:
Ah gotcha! That makes sense.

So in general, we prefer to have terraform based testing for terratest functions (integration tests) as opposed to having unit tests. There are 2 main advantages for this:

- We get to use terraform, a robust platform for managing infra, to do the test setup and teardown of resources that terratest depends on. This allows us to keep the API focused on validating infra, as opposed to managing infra.

- Users tend to refer to the examples for inspiration on testing. As such, having a terraform driven example + test, it leads to better discoverability of the functions (as opposed to a unit test which might not be as visible).

Reply:
this is super useful to understand - thanks for explaining! 👍